File: day10.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define N 64
#define A(i, j) a[(i) & (N - 1)][(j) & (N - 1)]
#define V(i, j) v[(i) & (N - 1)][(j) & (N - 1)]
char a[N][N];
bool v[N][N];
int n;

const int di[] = {-1, 0, 1, 0 };
const int dj[] = { 0, 1, 0,-1 };

int dfs(int i, int j, char target, bool track)
{
    if (A(i, j) != target) return 0;
    if (track) {
        if (V(i, j)) return 0;
        V(i, j) = true;
    }
    if (target == '9') return 1;

    int sum = 0;
    for (int d = 0; d < 4; d++) {
        sum += dfs(i + di[d], j + dj[d], target + 1, track);
    }
    return sum;
}

int main()
{
    FILE* file = fopen("day10.txt", "r");
    int sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%s", &A(n, 0))) n++;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            memset(v, 0, sizeof(v));
            sum += dfs(i, j, '0', true);
            sum2 += dfs(i, j, '0', false);
        }
    }

    printf("%d %d", sum, sum2);
}

File: day10.lua
#!/usr/bin/env lua

local mt = { __index = function(t, k) t[k] = {} return t[k] end }

local a = setmetatable({}, mt)
local v = setmetatable({}, mt)

local di = {-1, 0, 1, 0 }
local dj = { 0, 1, 0,-1 }

local zero = string.byte('0')
local nine = string.byte('9')

local function dfs(i, j, target, track)
    if a[i][j] ~= target then return 0 end
    if track then
        if v[i][j] then return 0 end
        v[i][j] = true
    end
    if target == nine then return 1 end

    local sum = 0
    for d = 1, 4 do
        sum = sum + dfs(i + di[d], j + dj[d], target + 1, track)
    end
    return sum
end

local file = io.open("day10.txt")
local sum = { 0, 0 }

for line in file:lines() do
    table.insert(a, { line:byte(1, #line) })
end

for i = 1, #a do
    for j = 1, #a do
        v = setmetatable({}, mt)
        sum[1] = sum[1] + dfs(i, j, zero, true)
        sum[2] = sum[2] + dfs(i, j, zero, false)
    end
end

print(sum[1], sum[2])
